### PR TITLE
CIP-0049? | ECDSA and Schnorr signatures in Plutus Core

### PR DESCRIPTION
**Pull request by kozross**
This CIP describes the ([recently merged](https://github.com/input-output-hk/plutus/pull/4368)) support for ECDSA and Schnorr signatures over the SECP256k1 curve. I've tried to include all the detail I can - any fields I am not sure about, I have left blank.

@michaelpj - does this seem like what you were looking for?

### CODE REVIEW

File: CIP-0042/README.md
Comment (michaelpj) [inline CIP-0042/README.md:105]: Let's make this "Path to active", and point out that the costing work will be done by the Plutus Core team.

**Review by michaelpj — APPROVED** [review]

File: CIP-0042/README.md
Comment (kozross) [reply CIP-0042/README.md]: Done.

File: CIP-0043/README.md
Comment (iquerejeta) [inline CIP-0043/README.md:122]:
```suggestion
  furthermore, if existing implementations have undergone review and audit, 
```

File: CIP-0042/README.md
Comment (iquerejeta) [inline CIP-0042/README.md:1]: Was this intentional?

File: CIP-0043/README.md
Comment (iquerejeta) [inline CIP-0043/README.md:1]: I think you are missing the `---` before the first line for this to parse correctly.

**Review by iquerejeta — COMMENTED** [review]
This looks good. I would, however, give a little more information on how the builtins work. I suggested that we add some further details on the serialisation format. I think it would also be useful to provide some detail on how the verification works. For Schnorr, it is sufficient to say that we are compatible with BIP340, and for ECDSA I would put somewhere more explicit that the function takes as input the hash of the message. It is important to insist that the verifier must hash the 'claimed' message before verification, and not accept the hash directly from the signer (maybe point to [this](https://bitcoin.stackexchange.com/a/81116/35586) explanation of why?)

File: CIP-0043/README.md
Comment (kozross) [reply CIP-0043/README.md:1]: Indeed I am! This is because of a dirty merge.

File: CIP-0043/README.md
Comment (kozross) [reply CIP-0043/README.md:83]: Agreed on the extra specification. However, if we're going to be specific, there's a few other things we need to mention:

* _x_ and _y_ coordinates of what?
* Is the _x_ coordinate a signed, or unsigned, integer?

I'm not sure what these are, and I'm sure the (mostly non-cryptographer) audience of this CIP probably won't know either.

File: CIP-0043/README.md
Comment (kozross) [reply CIP-0043/README.md:89]: As previously:

* What are `r` and `s` here?
* I assume they're integers: signed or unsigned?

File: CIP-0043/README.md
Comment (kozross) [reply CIP-0043/README.md:113]: I am fairly sure that it is the reference implementation for both. I agree about the history of use notes, will add those.

File: CIP-0043/README.md
Comment (iquerejeta) [reply CIP-0043/README.md]: I agree. I think we could give a brief intro on how ECDSA and Schnorr signatures are formed, and what constitutes a public key (or an EC point). These things would then be clear. I'm happy to write a draft myself and push it to the PR if you are OK with that.

Comment (kozross) [reply CIP-0043/README.md]: Go ahead!

File: CIP-0043/README.md
Comment (iquerejeta) [inline CIP-0043/README.md:93]: This bullet point and the next one might confuse the reader, as we say that a verification key "must correspond" to two different format. I would differentiate this bullet point with the next one using phrasing as follows: 
```
* A verification key corresponds to the _(x, y)_ coordinates of a point in the SECP256k1 curve, 
where _x, y_ are unsigned integers in big-endian form.
* The serialised (compressed) verification key format accepted by the builtins must correspond [...]
```

File: CIP-0043/README.md
Comment (iquerejeta) [inline CIP-0043/README.md:109]: Same as above. In this case it is a bit less confusing as both formats match, but maybe I would again rephrase in that the first bullet point is what the signature is, and the second is the compressed format accepted by the builtins.

File: CIP-0043/README.md
Comment (iquerejeta) [inline CIP-0043/README.md:121]: The same as above. Probably similar phrasing to that of the ECDSA verification keys would be good.

**Review by iquerejeta — COMMENTED** [review]
Looks good. Just some minor comments.

File: CIP-0043/README.md
Comment (KtorZ) [inline CIP-0043/README.md:91]: > Note that these differ from the serialization used by Bitcoin

Why is that? May we justify this choice under the _Rationale_ section?

File: CIP-0043/README.md
Comment (KtorZ) [inline CIP-0043/README.md:116]: Suggestion: some visual support to summarize the few points above. 

```
┏━━━━━━━━━━━━━━┯━━━━━━━━━━━━━━━┓
┃ r <32 bytes> │ s <32 bytes>  ┃
┗━━━━━━━━━━━━━━┷━━━━━━━━━━━━━━━┛
<--------- signature ---------->
```

File: CIP-0043/README.md
Comment (KtorZ) [inline CIP-0043/README.md:149]: This paragraph seems redundant with the _Motivation_ section.

File: CIP-0043/README.md
Comment (KtorZ) [inline CIP-0043/README.md:192]: :+1: cc @michaelpj, I assume you are aware of this.

File: CIP-0042/README.md
Comment (KtorZ) [reply CIP-0042/README.md:1]: Indeed, this seems like an error.

File: CIP-0043/README.md
Comment (KtorZ) [reply CIP-0043/README.md:2]: Please, rename the supporting folder and change the CIP number to `49` (without leading-zeros, otherwise Github markdown renderer has the brilliant idea to interpret those numbers in octal base :grimacing: ).

File: CIP-0043/README.md
Comment (michaelpj) [reply CIP-0043/README.md:192]: yep

**Comment by rphair** [comment]
@kozross this was on the CIP Editors' Meeting agenda today where it was suggested this proposal is being discussed, audited, or somehow moving forward in other channels.  If so, could you please update the status here so editors know how to move forward with this (or what we are waiting for)?

Also can you please renumber the CIP and its directory name as requested here (https://github.com/cardano-foundation/CIPs/pull/250#discussion_r963347317), so that we can merge this proposal when it becomes ready?

**Comment by iquerejeta** [comment]
The bindings for Schnorr and ECDSA signatures are currently being audited by bcryptic. We will soon have a report for it (ETA next week). The bindings have also been merged to cardano-base master (PRs [252](https://github.com/input-output-hk/cardano-base/pull/252), [258](https://github.com/input-output-hk/cardano-base/pull/258), [262](https://github.com/input-output-hk/cardano-base/pull/262/files) and [289](https://github.com/input-output-hk/cardano-base/pull/289)). I'll try to talk with Koz to modify the suggested changes.

**Comment by iquerejeta** [comment]
@rphair , @KtorZ , is there anything we can do to move this forward?

**Comment by rphair** [comment]
@iquerejeta apologies for the distracting references above (from someone else's CIP using your earlier assigned # 49).

We've put this on the agenda as "Last Check" for [CIP meeting # 56](https://discord.com/events/971785110770831360/1024688987245387806) (theme: "Plutus core changes") ([invite](https://discord.gg/qd6jE9Xj)) (2022-10-25 at 08:30 UTC).  It would be great if you and/or @kozross could be there: in any case it seems to me (@KtorZ?) like this will be merged unless there are any objections at the meeting.

**Comment by iquerejeta** [comment]
Great, I'll try to be there! 👍

**Review by KtorZ — APPROVED** [review]

**Pull request merged**
